File: 백준/Gold/1806. 부분합/부분합.py
import sys
input = sys.stdin.readline

n, s = map(int, input().split())
nums = list(map(int, input().split()))

pre_sum = [0]
temp = 0
for i in nums :
    temp += i
    pre_sum.append(temp)

#print(pre_sum)

start = 0
end = 0
INF = 1e9 + 100
result = INF

while(True) :
    val = pre_sum[end] - pre_sum[start]
    #print(val, result,start)
    if (val >= s) : # 조건 만족
        result = min(result, end-start)
        if (start < n) :
            start += 1 # 합을 더 작게
    else :
        if(end == n) : break
        # s 보다 크지도 않고.. val을 더이상 크게할수도 읎음
        if(end < n) :
            end += 1 # 합을 더 크게(val을 더크게)

if (result == INF) :
    print(0)
else :
    print(result)

File: 백준/Gold/1806. 부분합/부분합.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine(), " ");
		int n = Integer.parseInt(st.nextToken());
		int s = Integer.parseInt(st.nextToken());
		
		ArrayList<Integer> li = new ArrayList<>();
		int temp = 0;
		li.add(0);
		
		st = new StringTokenizer(br.readLine(), " ");
		for(int i=0; i<n; i++) {
			temp += Integer.parseInt(st.nextToken());
			li.add(temp);
		}
		// System.out.println(li);
		int val = 0;
		int INF = 9999999;
		int result = INF;
		int start = 0;
		int end = 0;
		while (true) {
			val = li.get(end) - li.get(start);
			
			if(val >= s) { 
				// 조건 만족
				result = Math.min(result, end-start);
				// 크기 줄이기
				if(start < li.size()-1 ) {
					start += 1;
				}
			} else {
				// 조건 X
				if(end == li.size()-1) {
					break;
				}
				if(end < li.size()-1) {
					end += 1;
				}
				
			}
		}
		if (result != INF) {
			System.out.println(result);
		} else {
			System.out.println("0");
		}
		
	}
}
